Prototype of much of the PEP

JelleZijlstra opened the pull request:
This isn't fully functional, but it should be enough to unblock work
on the runtime. I'll update this PR as I implement more.

DavidCEllis committed:
gh-103000: Optimise dataclasses asdict/astuple for common types (#103005)

Co-authored-by: Carl Meyer <carl@oddbird.net>
Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

CAM-Gerlach committed:
Doc: Avoid error lexing multiprocessing docs code block on Pygments 2.15.0 (#103421)

hugovk committed:
Docs: don't render files in includes/, they're for embedding only (#103313)

Zac-HD committed:
gh-99553: add tests for ExceptionGroup wrapping (#99615)

iritkatriel committed:
gh-102828: set stacklevel on deprecation warning (#103422)

iritkatriel committed:
gh-103176: sys._current_exceptions() returns mapping to exception instances instead of exc_info tuples (#103177)

iritkatriel committed:
gh-87092: move assembler related code from compile.c to assemble.c (#103277)

iritkatriel committed:
gh-91276: remove unused _PyOpcode_RelativeJump (#103156)

AlexWaygood committed:
gh-103373: `__mro_entries__` docs: improve cross references (#103398)

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

iritkatriel committed:
gh-77757: replace exception wrapping by PEP-678 notes in typeobject's __set_name__ (#103402)

sobolevn committed:
gh-87864: Use correct function definition syntax in the docs (#103312)

barneygale committed:
GH-103220: Fix `ntpath.join()` of partial UNC drive with trailing slash (GH-103221)

Zabolekar committed:
gh-103088: Sanitize venv paths when using MSYS or Cygwin Bash (GH-103325)

iritkatriel committed:
gh-87092: fix refleak in peepholer test harness (#103448)

benjaminp committed:
Remove redundant words from interpreter_definition.md. (GH-103455)

gaogaotiantian committed:
gh-103143: Polish pdb help messages and doc strings  (GH-103144)



* Made all the command part of the docstring match the official documentation
* Always have a space between the command and the description in docstring
* Added a helper function to format the help message

Before:

```
(Pdb) h a
a(rgs)
        Print the argument list of the current function.
(Pdb) h commands
commands [bpnumber]
        (com) ...
        (com) end
        (Pdb)
        ...
(Pdb) h interact
interact

        Start an interactive interpreter whose global namespace
        contains all the (global and local) names found in the current scope.
```

After
```
(Pdb) h a
      Usage: a(rgs)
      
      Print the argument list of the current function.
(Pdb) h commands
      Usage: (Pdb) commands [bpnumber]
             (com) ...
             (com) end
             (Pdb)
             ...
(Pdb) h interact
      Usage: interact
      
      Start an interactive interpreter whose global namespace
      contains all the (global and local) names found in the current scope.
```

Automerge-Triggered-By: GH:brandtbucher

gaogaotiantian committed:
gh-103237: Polish pdb docs (#103238)

bharel committed:
gh-103357: Add logging.Formatter defaults support to logging.config fileConfig and dictConfig (GH-103359)

furkanonder committed:
GH-83893: Cross reference env. vars and -X command line options (GH-103414)



Co-authored-by: Erlend E. Aasland

nburns committed:
gh-103417: use time.monotonic in the example for sched.scheduler (#103418)

aisk committed:
gh-103092: Isolate msvcrt (#103248)

markshannon committed:
GH-103082: Implementation of PEP 669: Low Impact Monitoring for CPython (GH-103083)

* The majority of the monitoring code is in instrumentation.c

* The new instrumentation bytecodes are in bytecodes.c

* legacy_tracing.c adapts the new API to the old sys.setrace and sys.setprofile APIs

erlend-aasland committed:
gh-103092: Isolate `_collections` (#103093)

Co-authored-by: Kumar Aditya <59607654+kumaraditya303@users.noreply.github.com>

sobolevn committed:
gh-103326: Remove `Python/importlib.h` (GH-103331)

Co-authored-by: Brett Cannon <brett@python.org>
Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

Zabolekar committed:
gh-103088: Fix virtual environment activate script not working in Cygwin (GH-103470)

smontanaro committed:
gh-67230: add quoting rules to csv module (GH-29469)



Add two quoting styles for csv dialects.
They will help to work with certain databases in particular.

Automerge-Triggered-By: GH:merwok

smontanaro committed:
gh-67230: document new csv quoting modes in whatsnew (gh-103491)

JelleZijlstra committed:
Run Tools/build/generate_global_objects.py

This fixes compilation for me locally.

JelleZijlstra committed:
Basic C implementation of TypeVar, TypeVarTuple, ParamSpec

This isn't fully functional, but it should be enough to unblock work
on the runtime. I'll update this PR as I implement more.

pradyunsg committed:
gh-95299: Rework test_cppext.py to not invoke setup.py directly (#103316)

* gh-95299: Rework test_cppext.py to not invoke setup.py directly

* Add tests/cppextdata data to `TESTSUBDIRS`

* Revert "Add tests/cppextdata data to `TESTSUBDIRS`"

This reverts commit 635492e53954fb0fc2a2875c8961bde99266c48d.

* Revert "gh-95299: Rework test_cppext.py to not invoke setup.py directly"

This reverts commit 41c5a667b5de7070bbde5780f1c124f96863c91d.

* Build and install the extension in a temporary directory instead

* Pull in wheels for setuptools and wheel for testing extension builds

alisaifee committed:
gh-103462: Ensure SelectorSocketTransport.writelines registers a writer when data is still pending (#103463)

tomasr8 committed:
gh-102978: Fix mock.patch function signatures for class and staticmethod decorators (#103228)

Fixes unittest.mock.patch not enforcing function signatures for methods
decorated with @classmethod or @staticmethod when patch is called with
autospec=True.

markshannon committed:
GH-103082: Turn on branch events for FOR_ITER instructions. (#103507)

Turn on branch events for FOR_ITER instructions.

Zabolekar committed:
gh-103088: Fix test_venv error message to avoid bytes/str warning (GH-103500)

markshannon committed:
GH-103488: Use return-offset, not yield-offset. (GH-103502)

* Use return-offset, not yield-offset, so that instruction pointer is correct when sending to a generator or coroutine.

ethanfurman committed:
gh-103365: [Enum] STRICT boundary corrections (GH-103494)

STRICT boundary:

- fix bitwise operations
- make default for Flag

ethanfurman committed:
gh-103479: [Enum] require __new__ to be considered a data type (GH-103495)

a mixin must either have a __new__ method, or be a dataclass, to be interpreted as a data-type

benjaminp committed:
Fix unused functions warnings in instrumentation.c (GH-103515)

JelleZijlstra committed:
fix TypeVar signature

JelleZijlstra committed:
fix TypeVar argument parsing

verhovsky committed:
Proofread howto/perf_profiling.rst (#103530)

JelleZijlstra committed:
ParamSpecArgs/Kwargs

JelleZijlstra committed:
make typing importable

JelleZijlstra committed:
Add some things

sobolevn committed:
gh-103406: Modernize pos-only arguments usage in `test_signature` (#103407)

aisk committed:
Fix syntax typo in isolating extensions doc (#103516)

kit1980 committed:
ipaddress: Remove non-existent ip_str param from docstring (#103461)

nedbat committed:
Remove double space in import error message (#103458)

sobolevn committed:
gh-103180: Add CI timeouts to all GitHub Actions jobs (#103437)

chrstphrchvz committed:
gh-103532: Remove TKINTER_PROTECT_LOADTK code (GH-103535)

This was only needed for Tk 8.4.13 and older,
but Tkinter already requires at least 8.5.12.

zware committed:
gh-103532: Add NEWS entry (#103542)

zware committed:
gh-103532: Fix reST syntax in NEWS entry (GH-103544)

erlend-aasland committed:
gh-103527: Add make deps for _codecs_* and _multibytecodec (#103528)

gaogaotiantian committed:
gh-101517: Add regression test for a lineno bug in try/except* impacting pdb (#103547)

barneygale committed:
GH-78079: Fix UNC device path root normalization in pathlib (GH-102003)

We no longer add a root to device paths such as `//./PhysicalDrive0`,
`//?/BootPartition` and `//./c:` while normalizing. We also avoid adding a
root to incomplete UNC share paths, like `//`, `//a` and `//a/`.

Co-authored-by: Eryk Sun <eryksun@gmail.com>

gvanrossum commented:
If you merge Eric's updated tests, do they pass? What's still missing?

JelleZijlstra commented:
I haven't looked at his tests yet (might spend some time on that tonight), but I don't think a lot of PEP 695-specific syntax will work until we get symtable changes implemented.

I was testing by just running `test_typing.py` (this branch replaces `typing.TypeVar` with the C version). I think there's ~140 failing tests; a lot around `*TypeVarTuple` not working, and a lot dealing with various aspects of TypeVar substitution.

gvanrossum commented:
Maybe we ought to focus on the symtable work first then? (In my experience the syntax *works* but has no semantics, but maybe I didn't explore deeply enough.)

JelleZijlstra committed:
Merge remote-tracking branch 'erictraut/type_param_syntax4' into tvobject

JelleZijlstra committed:
Make *Ts work

JelleZijlstra committed:
Just call typing.py

JelleZijlstra committed:
attempt at pickling

JelleZijlstra committed:
better approach

JelleZijlstra committed:
fix a few tests and refcount bugs

JelleZijlstra commented:
I got almost all tests in test_typing.py to pass by calling back into typing.py from the C implementation; the remaining ones are a little tedious but should be doable too. I'll start looking into the symtable now.

chgnrdv committed:
gh-102114: Make dis print more concise tracebacks for syntax errors in str inputs (#102115)

barneygale committed:
GH-103517: Improve tests for `pathlib.Path.walk()` (GH-103518)

sobolevn committed:
gh-103553: Improve `test_inspect`: add more assertions, remove unused (#103554)

erlend-aasland committed:
gh-103527: Add multibytecodec.h as make dep for _codecs_* (#103567)

cdce8p commented:
@JelleZijlstra FYI I've created #7 with the AST changes I suggested [here](https://discuss.python.org/t/pep-695-type-parameter-syntax/21646/99).

Eclips4 commented on Objects/typevarobject.c:
```diff
@@ -0,0 +1,813 @@
+                        "A single constraint is not allowed");
+        Py_XDECREF(bound);
+        return NULL;
+    } else if (n_constraints == 0) {
+        constraints = NULL;
+    } else if (bound != NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Constraints cannot be combined with bound=...");
+        Py_XDECREF(bound);
+        return NULL;
+    }
```
```suggestion
    }
    else if (n_constraints == 0) {
        constraints = NULL;
    } 
    else if (bound != NULL) {
        PyErr_SetString(PyExc_TypeError,
                        "Constraints cannot be combined with bound=...");
        Py_XDECREF(bound);
        return NULL;
    }
```
PEP 7

giampaolo committed:
gh-48330: address review comments to PR-12271 (#103209)

address review comments to PR-12271

Signed-off-by: Giampaolo Rodola <g.rodola@gmail.com>

erlend-aasland committed:
gh-103583: Add codecs and maps to _codecs_* module state (#103540)

ezio-melotti committed:
Remove `expert-*` from `project-updater` GH workflow (#103579)

corona10 committed:
gh-101525: Fix make test if the --enable-bolt enabled (gh-103574)

zooba committed:
gh-103088: Ensure POSIX venv scripts always use LF line endings (GH-103591)

Also touches the affected files in meaningless ways to ensure they get updated when pulling

aisk committed:
gh-103092: Isolate winreg (#103250)

sobolevn committed:
gh-103449: Fix a bug in dataclass docstring generation (#103454)

smontanaro committed:
gh-67230: update whatsnew note for csv changes (#103598)

rffontenelle committed:
GH-103484: Docs: add linkcheck allowed redirects entries for most cases (#103569)

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

pradyunsg committed:
gh-95299: Stop installing setuptools as a part of ensurepip and venv (#101039)

Remove the bundled setuptools wheel from ensurepip, and stop installing setuptools in environments created by venv.

Co-Authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>
Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>
Co-authored-by: Oleg Iarygin <oleg@arhadthedev.net>

Commit:
gh-101100: Fix broken reference `__format__` in `string.rst` (#103531)

zyckk4 committed:
[Doc] Fix a typo in optparse.rst (#103504)

zooba committed:
gh-95299: Remove lingering setuptools reference in installer scripts (GH-103613)

sobolevn committed:
GH-100530: Change the error message for non-class class patterns (GH-103576)

JelleZijlstra committed:
broken compiler before merge

JelleZijlstra committed:
Merge remote-tracking branch 'erictraut/type_param_syntax4' into tvobject

JelleZijlstra committed:
Works for functions

JelleZijlstra committed:
fall back to normal scope

ethanfurman committed:
gh-103596: [Enum] do not shadow mixed-in methods/attributes (GH-103600)

For example:

    class Book(StrEnum):
        title = auto()
        author = auto()
        desc = auto()

    Book.author.desc is Book.desc

but

    Book.author.title() == 'Author'

is commonly expected.  Using upper-case member names avoids this confusion and possible performance impacts.

Co-authored-by: samypr100 <3933065+samypr100@users.noreply.github.com>

erlend-aasland committed:
gh-103617: Fix compiler warning in _iomodule.c (#103618)

erlend-aasland committed:
gh-103583: Always pass multibyte codec structs as const (#103588)

sobolevn committed:
gh-103582: Remove last references to `argparse.REMAINDER` from docs (#103586)

iritkatriel committed:
gh-102778: IDLE - make sys.last_exc available in Shell after traceback (#103314)

---------
Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

kumaraditya303 committed:
GH-88342: clarify that `asyncio.as_completed` accepts generators yielding tasks (#103626)

erlend-aasland committed:
gh-83004: Harden msvcrt further (#103420)

JelleZijlstra committed:
does not exactly work

erlend-aasland committed:
gh-103583: Add ref. dependency between multibytecodec modules (#103589)

pablogsal committed:
gh-102856: Initial implementation of PEP 701 (#102855)

Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>
Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>
Co-authored-by: Marta Gómez Macías <mgmacias@google.com>
Co-authored-by: sunmy2019 <59365878+sunmy2019@users.noreply.github.com>

pablogsal committed:
gh-102856: Skip test_mismatched_parens in WASI builds (#103633)

pablogsal committed:
gh-102856: Clean some of the PEP 701 tokenizer implementation (#103634)

abalkin committed:
gh-83861: Fix datetime.astimezone() method (GH-101545)

ericsnowcurrently committed:
gh-98040: Move the Single-Phase Init Tests Out of test_imp (gh-102561)

I recently added some tests to test_imp, but @warsaw is removing that file in gh-98573. The tests are worth keeping so here I'm moving them to test_import.

why-not-try-calmer committed:
gh-98641: Document difference between task group and gather (#103644)

The purpose of the comments is to rule out the implication that asyncio.TaskGroup is a drop-in replacement / better alternative to asyncio.gather().

wheelerlaw committed:
gh-99352: Respect `http.client.HTTPConnection.debuglevel` in `urllib.request.AbstractHTTPHandler`  (#99353)

* bugfix: let the HTTP- and HTTPSHandlers respect the value of http.client.HTTPConnection.debuglevel

* add tests

* add news

* ReSTify NEWS and reword a bit.

* Address Review Comments.

* Use mock.patch.object instead of settting the module level value.
* Used test values to assert the debuglevel.

---------

Co-authored-by: Gregory P. Smith <greg@krypto.org>
Co-authored-by: Senthil Kumaran <senthil@python.org>

jaraco committed:
Sync with importlib_metadata 6.5 (GH-103584)

JelleZijlstra committed:
commit before new approach

JelleZijlstra committed:
New approach

arhadthedev committed:
gh-103082: Fix shifted field initialization in `instrumentation.c` (GH-103561)

Fix shifted field initialization in instrumentation.c

JelleZijlstra committed:
correctly make it a cellvar

jaraco committed:
gh-103661: Skip failing test on Windows. (#103662)

JelleZijlstra committed:
Now it works

JelleZijlstra committed:
fix defaults

JelleZijlstra committed:
undo some unnecessary changes

JelleZijlstra committed:
support async def

JelleZijlstra committed:
start class support

JelleZijlstra committed:
detect duplicate type params

rhettinger committed:
GH-103415: Document itertools.batched() in whatsnew.3.12 (#103670)

rhettinger committed:
Minor improvements to the functools docs (#103672)

* Use an f-string for improved readability

* Put version notes in chronological order

JelleZijlstra committed:
clean up some whitespace

sobolevn committed:
gh-103556: [inspect.Signature] disallow pos-or-kw params without default after pos-only with default (#103557)

JelleZijlstra committed:
Improve some tests

JelleZijlstra committed:
No runtime error there

JelleZijlstra committed:
actually generate a typevar

pfmoore committed:
gh-103559: Update bundled pip version to 23.1.1 (gh-103560)

* Update bundled pip version to 23.1
* Update to pip 23.1.1

JelleZijlstra committed:
refactor typeparam blocks

rhettinger committed:
Descriptor HowTo:  Update to include attributes added in Python 3.10  (GH-103666)

rhettinger committed:
GH-103475: cache() and lru_cache() do not have a "call once" guarantee (GH-103669)

rffontenelle committed:
GH-103484: Fix broken links reported by linkcheck (#103608)

* Doc: Fix broken links reported by linkcheck

* Apply suggestions from code review

- Remove extra diff line in faq/library.rst (merwok)
- Use HTTPS to link Unicode 15.0.0 to solve a redirect (hugovk)
- Use wayback machine link for openssl 1.1.0 instead of linking 1.1.1, "as this text mentions a feature from 1.1.0" (hugovk)

Co-authored-by: Éric <merwok@netwok.org>
Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

* Doc: Make mark-up code as literal

* Doc: Alphabetize items in linkcheck_ignore

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

* Doc: Improve comment in sphinx conf

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

---------

Co-authored-by: Éric <merwok@netwok.org>
Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>
Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

JelleZijlstra committed:
Fix nested class scopes (functions)

JelleZijlstra committed:
fix it in classes too

JelleZijlstra committed:
Support bounds

eric-wieser committed:
gh-83791: Raise TypeError for len(memoryview_0d) (#18463)

Changes the behaviour of `len` on a zero-dimensional `memoryview` to raise `TypeError`. Previously, `len` would return `1`.

JelleZijlstra committed:
Use dummy class to get TypeVar-like types

JelleZijlstra committed:
fix typo in test

JelleZijlstra committed:
Add union support to TypeVar (and ParamSpec, somewhat dubiously)

jaraco committed:
gh-103661: Apply bugfix from importlib_metadata 6.5.1 and restore test. (#103681)

eduardo-elizondo committed:
gh-84436: Implement Immortal Objects (gh-19474)

This is the implementation of PEP683

Motivation:

The PR introduces the ability to immortalize instances in CPython which bypasses reference counting. Tagging objects as immortal allows up to skip certain operations when we know that the object will be around for the entire execution of the runtime.

Note that this by itself will bring a performance regression to the runtime due to the extra reference count checks. However, this brings the ability of having truly immutable objects that are useful in other contexts such as immutable data sharing between sub-interpreters.

hugovk committed:
Revert "Avoid error lexing multiprocessing docs code block on Pygments 2.15.0" (#103616)

This reverts commit ace51dcdb781b0608b1273d246ebaee849561435.

san-juan1667 committed:
gh-68654: Clarify subdirectories used by pkgutil.extend_path (#103701)

Clarify sub directories used by pkgutil.extend_path in the docs and the docstring

sobolevn committed:
gh-102310: Change error range for invalid bytes literals (#103663)

JelleZijlstra committed:
add function.__type_variables__

hemberger committed:
gh-81403: Fix for CacheFTPHandler in urllib (#13951)

bpo-37222: Fix for CacheFTPHandler in urllib

A call to FTP.ntransfercmd must be followed by FTP.voidresp to clear
the "end transfer" message. Without this, the client and server get
out of sync, which will result in an error if the FTP instance is
reused to open a second URL. This scenario occurs for even the most
basic usage of CacheFTPHandler.

Reverts the patch merged as a resolution to bpo-16270 and adds a test
case for the CacheFTPHandler in test_urllib2net.py.

Co-authored-by: Senthil Kumaran <senthil@python.org>

adriangb committed:
GH-103699:  Add `__orig_bases__` to various typing classes (#103698)

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>
Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

JelleZijlstra committed:
Generic work

JelleZijlstra committed:
Use _Py_DECLARE_STR

corona10 committed:
gh-101408: PyObject_GC_Resize should calculate preheader size. (gh-101741)

itsankitkp committed:
gh-103716: Add test support requires fork in simple_subprocess  (gh-103717)

add requires fork as test case depends on this

sobolevn committed:
gh-103592: Add tests of `Literal` with `Enum` and `Union` of `Literal`s (#103706)

lysnikolaou committed:
GH-103718: Correctly cache and restore f-string buffers when needed (GH-103719)

JelleZijlstra committed:
Support inheriting from Generic

Gobot1234 committed:
gh-101688: Implement types.get_original_bases (#101827)

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

JelleZijlstra committed:
Fix up qualnames

JelleZijlstra committed:
Update the magic number, rearrange typing.py accordingly

JelleZijlstra committed:
fix missing initialization

JelleZijlstra committed:
set __type_variables__ on classes

JelleZijlstra committed:
Add docs to typevarobject.c

JelleZijlstra committed:
drop Python implementations of Generic, TypeVar, etc.

JelleZijlstra committed:
fix Generic

JelleZijlstra committed:
make generics instantiable

JelleZijlstra committed:
fix NamedTuple

JelleZijlstra committed:
fix some tests

JelleZijlstra committed:
one more

orsenthil committed:
gh-81403: Add NEWS entry for PR #13951 (#103729)

Add NEWS entry for PR #13951

itsankitkp committed:
gh-103724: Add test case if no arg as provided in os.register_at_fork (gh-103725)

sobolevn committed:
gh-103668: Run pyugrade on idlelib (#103671)


---------
Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

chrstphrchvz committed:
gh-103685: Fix tkinter.Menu.index() for Tk 8.7 (#103686)

---------

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>
Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>

gaogaotiantian committed:
gh-103285: Rewrite _splitlines_no_ff to improve performance (#103307)

encukou committed:
gh-102950: Implement PEP 706 – Filter for tarfile.extractall (#102953)

JelleZijlstra committed:
Use a heap type for Generic

JelleZijlstra committed:
Better way to set __type_variables__ in the class

JelleZijlstra committed:
Fix TypeVar | "ForwardRef"

chrstphrchvz committed:
Change 'dependant' to 'dependent' (#103745)

The word 'dependent' is both an adjective and a noun. A 'dependant' is a British alternative spelling for the noun form. In idlelib.sidebar, 'OS-dependant' is an adjective and clearly wrong. In 'Using', 'dependant' as a noun would be acceptable in Britain, but we use American spellings in Python docs.
https://www.merriam-webster.com/words-at-play/spelling-variants-dependent-vs-dependant

slateny committed:
gh-44123: Add note on relative path for os.exec* (GH-93826)

Co-authored-by: Steve Dower <steve.dower@python.org>

alex committed:
gh-103712: Increase the length of the type name in AttributeError messages (#103713)

Eclips4 committed:
Add tests for empty range equality (#103751)

JelleZijlstra committed:
Make ParamSpecArgs and Kwargs not subclassable

JelleZijlstra committed:
Correctly unpack TypeVarTuple

swtaarrs committed:
gh-103091: Add PyUnstable_Type_AssignVersionTag (#103095)

hroncok committed:
GH-102973: Slim down Fedora packages in the dev container (#103283)

richard-hajek committed:
Removed unnecessary escaping of asterisks (#103714)

Removed unnecessary escaping of asterisks, as visible both on GitHub and in https://docs.python.org/3/library/decimal.html#decimal.localcontext

Co-authored-by: Łukasz Langa <lukasz@langa.pl>

JelleZijlstra committed:
Add support for type aliases (but not for computing the value)

zitterbewegung committed:
gh-87452: Improve the Popen.returncode docs

Clarifies that it remains None until a method checking the child process status has been called and noticed it has terminated.

jaimealonso committed:
gh-103765: Fix 'Warning: py:class reference target not found: ModuleSpec' (GH-103769)

sobolevn committed:
gh-103746: Test `types.UnionType` and `Literal` types together (#103747)

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into tvobject

JelleZijlstra committed:
fix merge

alago1 committed:
gh-91687: modernize dataclass example typing (#103773)

modernize dataclass example typing `list` rather than `List` and comment as to that line being the alluded too error.

ericsnowcurrently committed:
gh-103323: Get the "Current" Thread State from a Thread-Local Variable (gh-103324)

We replace _PyRuntime.tstate_current with a thread-local variable. As part of this change, we add a _Py_thread_local macro in pyport.h (only for the core runtime) to smooth out the compiler differences. The main motivation here is in support of a per-interpreter GIL, but this change also provides some performance improvement opportunities.

Note that we do not provide a fallback to the thread-local, either falling back to the old tstate_current or to thread-specific storage (PyThread_tss_*()). If that proves problematic then we can circle back. I consider it unlikely, but will run the buildbots to double-check.

Also note that this does not change any of the code related to the GILState API, where it uses a thread state stored in thread-specific storage. I suspect we can combine that with _Py_tss_tstate (from here). However, that can be addressed separately and is not urgent (nor critical).

(While this change was mostly done independently, I did take some inspiration from earlier (~2020) work by @markshannon (main...markshannon:threadstate_in_tls) and @vstinner (#23976).)

furkanonder committed:
GH-65022: Fix description of copyreg.pickle function (#102656)

Fix description of copyreg.pickle function

sunmy2019 commented:
[6312d7b](https://github.com/erictraut/cpython/pull/3/commits/6312d7bdd84f6821a307b4a09156d885a6230f2d) got test failure (segment fault) for `test_compile`.

By the way, why is this PR so huge? Shouldn't it be a series of small PRs that merged to [erictraut:type_param_syntax4](https://github.com/erictraut/cpython/tree/type_param_syntax4) somehow?

arhadthedev committed:
gh-94518: Port 23-argument `_posixsubprocess.fork_exec` to Argument Clinic (#94519)

Convert fork_exec to pre-inlined-argparser Argument Clinic

Co-authored-by: Gregory P. Smith <greg@krypto.org>

terryjreedy committed:
Revert "Add tests for empty range equality (#103751)" (#103770)

This reverts commit dca27a69a8261353f7f986eb8f808f0d487ac4b7.
Added tests are redundant with existing tests.

lysnikolaou committed:
GH-103727: Avoid advancing tokenizer too far in f-string mode (GH-103775)

iritkatriel committed:
gh-87092: fix a few cases of incorrect error handling in compiler (#103456)

gh-87092: fix a few cases of incorrect error handling

zware committed:
gh-103776: Remove explicit uses of $(SHELL) from Makefile (GH-103778)

This avoids conflicting with the shebang of the called scripts as well
as avoiding hard errors on platforms where the called script runs a
failing unchecked command in the usual course of checking since
`SHELL=/bin/sh -e` as of a90863c.

Fixes gh-103776.

sunmy2019 commented:
> [6312d7b](https://github.com/erictraut/cpython/pull/3/commits/6312d7bdd84f6821a307b4a09156d885a6230f2d) got test failure (segment fault) for `test_compile`.

Minimal case
```py
import ast

_ast = compile("class _: pass", "2", "exec", ast.PyCF_ONLY_AST)
compile(_ast, "", "exec")
```

howiezhao committed:
gh-94300: Update datetime.strptime documentation (#95318)

The new wording better reflects the cases where `datetime.strptime` differs from` time.strptime`.

---------

Co-authored-by: Paul Ganssle <git@m.ganssle.io>

sunmy2019 commented on Python/compile.c:
```diff
@@ -2119,6 +2058,48 @@ wrap_in_stopiteration_handler(struct compiler *c)
+        return SUCCESS;
+    }
+    Py_ssize_t n = asdl_seq_LEN(typeparams);
+
```
The newest crash happens if `n==0`. Not guarded here.

I guess 0 should not be allowed here, somewhere else is broken. Also, an assert should be added here.

JelleZijlstra replied:
Thanks, I thought it would be NULL if there are no type params but apparently that's not true. I pushed a fix.

wjandrea committed:
gh-99032: datetime docs: Encoding is no longer relevant (#93365)

This removes a section of the `strftime` and `strptime` documentation that refers to a bygone era when `strftime` would return an encoded byte string.

---------

Co-authored-by: Paul Ganssle <1377457+pganssle@users.noreply.github.com>

sobolevn committed:
gh-99184: Bypass instance attribute access in `repr` of `weakref.ref` (#99244)

JelleZijlstra committed:
Avoid complicating compile_call_helper

JosephSBoyle committed:
gh-102498 Clean up unused variables and imports in the email module  (#102482)

* Clean up unused variables and imports in the email module

* Remove extra newline char

* Remove superflous dict+unpacking syntax

* Remove unused 'msg' var

* Clean up unused variables and imports in the email module

* Remove extra newline char

* Remove superflous dict+unpacking syntax

* Remove unused 'msg' var

---------

Co-authored-by: Barry Warsaw <barry@python.org>

JelleZijlstra committed:
Apparently it is a length-0 list if there are no typeparams, not NULL

JelleZijlstra commented:
> [6312d7b](https://github.com/erictraut/cpython/pull/3/commits/6312d7bdd84f6821a307b4a09156d885a6230f2d) got test failure (segment fault) for `test_compile`.
> 
> By the way, why is this PR so huge? Shouldn't it be a series of small PRs that merged to [erictraut:type_param_syntax4](https://github.com/erictraut/cpython/tree/type_param_syntax4) somehow?

I can't merge into Eric's branch, and by now there's enough for a reasonably complete CPython PR. Unfortunately it's hard to split up the change as there are few self-contained subparts.

erictraut commented:
@JelleZijlstra, would you like me to merge into my branch now? I could also create a new branch (e.g. `erictraut:type_param_syntax5`) that rebases all of these changes onto the latest cpython main branch. That would make it easier to tell what you changed.

erictraut commented:
Or would you prefer to submit the PR to cpython directly from your branch?

JelleZijlstra commented:
I already submitted a draft PR with my branch, I think that's a good outcome. Your parser changes seem solid, I haven't needed to change anything there.

I talked to Thomas about the change I suggested in #11 and he's receptive to it, but we will need SC approval. I can handle the communication for that. I'm working with various other core devs too to get advice.

erictraut commented:
Sounds good. Thanks for all of your work on this! Looks like you've made great progress.

sunmy2019 commented:
I will keep working on reviewing this in the next couple of weeks.

I will move to that PR.

JelleZijlstra committed:
Fix GC for TypeAlias

JelleZijlstra committed:
fix warnings

JelleZijlstra committed:
Add to globals-to-fix

JelleZijlstra committed:
Add to _freeze_module.vcxproj

itamaro committed:
gh-103780: Use patch instead of mock in asyncio unix events test (#103782)

iritkatriel committed:
gh-101517: fix line number propagation in code generated for except* (#103550)

hugovk committed:
gh-101100: Fix Sphinx warnings in `argparse` module (#103289)

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>
Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

hauntsaninja committed:
gh-103492: Clarify SyntaxWarning with literal comparison (#103493)

ericsnowcurrently committed:
gh-100227: Only Use deepfreeze for the Main Interpreter (gh-103794)

Deep-frozen code objects are cannot be shared (currently) by
interpreters, due to how adaptive specialization can modify the
bytecodes. We work around this by only using the deep-frozen objects in
the main interpreter. This does incur a performance penalty for
subinterpreters, which we may be able to resolve later.

ambv committed:
gh-103791: Make contextlib.suppress also act on exceptions within an ExceptionGroup (#103792)

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

carljm committed:
gh-87729: add LOAD_SUPER_ATTR instruction for faster super() (#103497)

This speeds up `super()` (by around 85%, for a simple one-level
`super().meth()` microbenchmark) by avoiding allocation of a new
single-use `super()` object on each use.

ericsnowcurrently committed:
gh-95795: Move types.next_version_tag to PyInterpreterState (gh-102343)

Core static types will continue to use the global value.  All other types
will use the per-interpreter value.  They all share the same range, where
the global types use values < 2^16 and each interpreter uses values
higher than that.

jb2170 committed:
gh-103673: Add missing ForkingUnixStreamServer and ForkingUnixDatagramServer socketservers (#103674)

sockserver gains ForkingUnixStreamServer and ForkingUnixDatagramServer classes for consistency with all of the others. Ironically these existed but were buried in our test suite.

Addresses #103673 

<!-- gh-issue-number: gh-103673 -->
* Issue: gh-103673
<!-- /gh-issue-number -->

---------

Co-authored-by: Gregory P. Smith <greg@krypto.org>
Co-authored-by: Nikita Sobolev <mail@sobolevn.me>

dversoza committed:
gh-103801: Tools/wasm linting and formatting (#103796)

This PR makes three minor linting adjustments to the `wasm` module
caught by [ruff](https://github.com/charliermarsh/ruff).

<!-- gh-issue-number: gh-103801 -->
* Issue: gh-103801
<!-- /gh-issue-number -->

---------

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>

JelleZijlstra committed:
Use heap types, fixing pickling

JelleZijlstra committed:
Fix GC

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into tvobject

ericsnowcurrently committed:
gh-101659: Isolate "obmalloc" State to Each Interpreter (gh-101660)

This is strictly about moving the "obmalloc" runtime state from
`_PyRuntimeState` to `PyInterpreterState`.  Doing so improves isolation
between interpreters, specifically most of the memory (incl. objects)
allocated for each interpreter's use.  This is important for a
per-interpreter GIL, but such isolation is valuable even without it.

FWIW, a per-interpreter obmalloc is the proverbial
canary-in-the-coalmine when it comes to the isolation of objects between
interpreters.  Any object that leaks (unintentionally) to another
interpreter is highly likely to cause a crash (on debug builds at
least).  That's a useful thing to know, relative to interpreter
isolation.

rodolfomp123 committed:
GH-103805: Lib test f541 linting issue fix (#103812)

This PR makes some minor linting adjustments to the Lib/test module
caught by [ruff](https://github.com/charliermarsh/ruff). The adjustments
are all related to the `F541 f-string without any placeholders` issue.

Issue: https://github.com/python/cpython/issues/103805

<!-- gh-issue-number: gh-103805 -->
* Issue: gh-103805
<!-- /gh-issue-number -->

---------

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Gregory P. Smith <greg@krypto.org>

olgarithms committed:
gh-103810: Fix broken references in dataclasses (#103811)

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

jaimealonso committed:
gh-103765: Remove pkgutil.rst from .nitignore (GH-103797)

ambv committed:
Also fail DO-NOT-MERGE when "awaiting changes" or "awaiting change review" present on PR (#103807)

"awaiting changes" means somebody put a review that requested changes.

"awaiting change review" means that the PR author published changes
after a red review and then requested a re-review.

JelleZijlstra committed:
Self-review

- Use the right name for the new TypeVar attribute (infer_variance, not autovariance)
- Simpler intrinsics (avoid some wrapper functions)
- Undo some unnecessary changes
- Consistent styling of function declarations

JelleZijlstra committed:
It's supposed to be called TypeAliasType

alago1 committed:
gh-82814: fix shutil access error on WSL (#103790)

gh-82814: Adds `errno.EACCES` to the list of ignored errors on
`_copyxattr`.  EPERM and EACCES are different constants but
in general should be treated the same.

News entry authored by: Gregory P. Smith <greg@krypto.org>

JelleZijlstra committed:
Allow evaluating TypeAlias

lysnikolaou committed:
gh-103718: Correctly set f-string buffers in all cases (GH-103815)

Turns out we always need to remember/restore fstring buffers in all of
the stack of tokenizer modes, cause they might change to
`TOK_REGULAR_MODE` and have newlines inside the braces (which is when we
need to reallocate the buffer and restore the fstring ones).

Agent-Hellboy committed:
gh-103636: add enums for days and months in calendar module (GH-103642)

Convert day contants to a Day enum, and add a Month enum.

ericsnowcurrently committed:
gh-100227: Add a Granular Lock for _PyRuntime.imports.extensions.dict (gh-103460)

The lock is unnecessary as long as there's a GIL, but completely
necessary with a per-interpreter GIL.

JelleZijlstra committed:
Support __or__ on type aliases

JelleZijlstra committed:
fix | and subscripting on TypeAliasType

JelleZijlstra committed:
Test fixes and new tests

sunmy2019 committed:
gh-103826: fix unused variable warning introduced in gh-102343 (#103825)

partev committed:
DOC: remove reference to OpenSolaris (#102671)

JelleZijlstra committed:
Add syntax error for nonlocal as specified by the PEP

corona10 committed:
gh-103726: Set up gcc-10 for ASAN (gh-103728)

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a
typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-103726 -->
* Issue: gh-103726
<!-- /gh-issue-number -->

JelleZijlstra committed:
Implement TypeAlias repr

JelleZijlstra committed:
__type_variables__ -> __type_params__

JelleZijlstra committed:
Remove POP_NULL

encukou committed:
gh-102950: Adjust tarfile filter tests for systems that don't set the sticky bit (GH-103831)

Also remove expilcit `type=tarfile.DIRTYPE`, the slash at the end is
enough.

JelleZijlstra committed:
Lazily evaluate TypeVar bounds/constraints

JelleZijlstra committed:
Add NEWS

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into tvobject

JelleZijlstra committed:
Disallow yield/yield from/await/walrus

JelleZijlstra committed:
fix doctests

JelleZijlstra committed:
Stop altering the execution environment

kj7rrv committed:
gh-51574: Make tempfile.mkdtemp() always return absolute paths (#94612)

Co-authored-by: Éric <merwok@netwok.org>
Co-authored-by: AlexWaygood <alex.waygood@gmail.com>

JelleZijlstra committed:
Disallow walrus/yield while evaluating generics

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into tvobject

AlexWaygood committed:
gh-103721: Improve cross-references for generic-alias docs (#103838)

Cc. @adriangb

The "stub documentation" in `types.rst` does already link to the
in-depth docs in `stdtypes.rst`, but the link isn't obvious for new
users. It deserves to be made more prominent.

- Issue: https://github.com/python/cpython/issues/103721

carljm committed:
gh-87729: specialize LOAD_SUPER_ATTR_METHOD (#103809)

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into tvobject

JelleZijlstra committed:
regen all, minor quality improvements

JelleZijlstra committed:
fix leaked ref to the TypeAlias type

JelleZijlstra committed:
Fix another leak

gvanrossum commented:
I propose that Eric close this in favor of https://github.com/python/cpython/pull/103764

erictraut commented:
Closed.
